obsolete/weather/weather/wm918/wx200d-1.2/wx200d.5
.\" @(#)$Id: wx200d.5,v 1.1 2002/09/23 19:12:51 bogdan Rel $
.\"
.\" Copyright (C) 1998 - 2001 Tim Witham <twitham@quiknet.com>
.\"
.\" (see the files README and COPYING for more details)

.TH WX200D 5 "Jun 02, 2001" "wx200d Manual"
.SH NAME
wx200d \- format of log files created by wx200d
.SH SYNOPSIS
.B ${prefix}/share/wx200d/YYYYMMDD.bin[.gz]
.B ${prefix}/share/wx200d/YYYYMMDD.tab[.gz]
.B ${prefix}/share/wx200d/YYYYMMDD.day
.B ${prefix}/share/wx200d/YYYYMM.mon

.SH DESCRIPTION

The files in ${prefix}/share/wx200d are WX200 weather data logs
created by
.B wx200d(8)
or wxsum.

The base filename is made up of the year, month and day of the data
contained in the file.  This is of the form YYYYMMDD which makes it
easy to sort the files in time order.  The extension of the file
indicates the format of the content and can currently be one of the
following formats.  Of course a
.I .gz
suffix indicates the file has been compressed with
.B gzip(1).
.P

.TB 0.5i
.B YYYYMMDD.bin[.gz]

These files contain "abbreviated" binary WX200 data.  WX200 data
"groups" (see wx200.txt for details) are logged directly from the
device but only when they have changed.  This results in the smallest
possible file that can still be used by programs that understand the
complete WX200 serial data stream.
.P

This abbreviation is not lossy.  That is, all bits of data are
recorded with no loss of information.  These
.I .bin
files can therefore be used to create
.I .tab
files, via
.B wxfilter(1)
, but not vice-versa.
.P

.TB 0.5i
.B YYYYMMDD.tab[.gz]

These files contain a tab-delimited ASCII log of the WX200 sensor
values.  They may be conveniently plotted by
.B wxplot(1)
or
.B gnuplot(1)
or imported into spreadsheets or other systems.
.P

When any variable changes, a whole line is recorded.  The columns of
the line are delimited by a single tab character and record the
following WX200 sensor values in order:

.TB 0.5i
.B Time
The hour of the day this line of data occurred.  The integer part is
the hour in the range 0 to 23.  The fractional part is the fraction of
the hour.  The time is recorded in this way so that this field can be
plotted linearly on the X axis of a graph.

.TB 0.5i
.B indoor temperature
0.0 to 50.0 degrees Celsius @ 0.1

.TB 0.5i
.B outdoor temperature
-40.0 to 60.0 degrees Celsius @ 0.1

.TB 0.5i
.B indoor humidity
10 to 97 percent @ 1

.TB 0.5i
.B outdoor humidity
10 to 97 percent @ 1

.TB 0.5i
.B indoor dew point
0 to 47 degrees Celsius @ 1

.TB 0.5i
.B outdoor dew point
0 to 56 degrees Celsius @ 1

.TB 0.5i
.B local barometric pressure
795 to 1050 millibars @ 1

.TB 0.5i
.B sea level pressure
795.0 to 1050.0 millibars @ 0.1

.TB 0.5i
.B gust wind direction
0 to 350 compass degrees @ 10

.TB 0.5i
.B gust wind speed
0.0 to 56.0 meters per second @ 0.2

.TB 0.5i
.B average wind direction
0 to 350 compass degrees @ 10

.TB 0.5i
.B average wind speed
0.0 to 56.0 meters per second @ 0.2

.TB 0.5i
.B wind chill
-85 to 60 degrees Celsius @ 1

.TB 0.5i
.B rainfall rate
0 to 998 millimeters per hour @ 1

.TB 0.5i
.B cumulative rainfall since reset
0 to 9999 millimeters @ 1

.TB 0.5i
.B YYYYMMDD.day

These daily summary files contain hourly average, low and high for
each variable, one line per hour.  The file format is very similar to
the
.I .tab
files described above.  The difference is that after the hour
timestamp, the group of data columns appears three times to record the
average, low and high of each variable in that order.  The wind
direction is the only exception to this: the three most common wind
directions are listed in order.

.TB 0.5i
.B YYYYMM.mon

These monthly summary files contain daily average, low and high for
each variable, one line per day.  The file format is just like the
.I .day
files described above except that the first column is in the YYYYMMDD
format to indicate the date of the data instead of the hour.


.SH FILES
${prefix}/share/wx200d/YYYYMMDD.bin[.gz]
${prefix}/share/wx200d/YYYYMMDD.tab[.gz]
${prefix}/share/wx200d/YYYYMMDD.day
${prefix}/share/wx200d/YYYYMM.mon

.SH LIMITATIONS

Files created by versions less than 1.0 lacked the rain total column.
The wxsum script can correct this for
.I .day
and
.I .mon
summary files if you still have the original
.I .tab
files available.  Otherwise, old files will not be compatible with
.B wxplot(1).

.SH SEE ALSO
wx200d(8), wxplot(1), gnuplot(1), wxfilter(1), wxdebug(1)

.SH AUTHOR
Tim Witham <twitham@quiknet.com>
.SH COPYRIGHT
Copyright (C) 1998 - 2001 Tim Witham <twitham@quiknet.com>

wx200d is released under the conditions of the GNU General Public
License.  See the files README and COPYING in the distribution for
details.


sip_router/NEWS
XXX content of this file reflects a future release;
    features described here may or may not be available
    in the CVS tree
    

Release notes for SIP Express Router (ser)
***********************************************

$Id: NEWS,v 1.12 2003/04/18 14:11:30 jiri Exp $

***********************************************
* Changes introduced in 0.8.11
***********************************************

+--------------------------------------------------------+
| CAUTION: the 0.8.11 release include changes which      |
| are incompatible with scripts and databases used       |
| in previous versions. Care is advised when upgrading   |
| from previous releases to 0.8.11.                      |
+--------------------------------------------------------+

New features
=============
- RFC3261 support
	- TCP support and cross-transport forwarding [core]
	- loose routing support [rr module]
- New modules
	- vm -- voicemail interface [vm]
	- ENUM support [enum]
	- presence agent [pa]
	- dynamic domain management -- allows to manipulate 
  	  hosting of multiple domains in run-time [module]
	- flat-text-file database support [dbtext]
	- rich access control lists [permissions]
- Feature Improvements
	- click-to-dial, which is based on improved tm/FIFO 
	  that better supports external applications [tm module]
	- web accounting -- acc module can report to serweb
  	  on placed calls [acc module]
	- improved exec module (header fields passed now
      as environment variables to scripts) [exec module]
- Architectural Improvements
	- powerpc fast locking support
	- netbsd support
	- 64 bits arch. support (e.g. netbsd/sparc64).
- New Experimental Features (not tested at all yet)
	- nathelper utility for Cisco/ATA NAT traversal [nathelper]
	- another NAT traversal utility [mangler]
	- postgress support [postgress]
	- fcp module [fcp]
	- pdt module (prefix2domain) [pdt]

Changes to use of ser scripts
=============================

About Multiple Transport Support
--------------------------------
SER now suports multiple transport protocols: UDP and TCP. As there
may be UAs which support only either protocol and cannot speak to
each other directly, we recommend to alway record-route SIP requests,
to keep the transport-translating SER in path. Also, if a destination
transport is not known, stateful forwarding is recommended -- use of
stateless forwarding for TCP2UDP would result in loss of reliability.


core
----
- reply_route has been renamed to failure_route -- the old name caused
  too much confusion
- forward_tcp and forward_udp can force SER to forward via specific
  transport protocol

acc module:
-----------
- radius and sql support integrated in this module; you need to
  recompile to enable it
- acc_flag is now called log_flag to better reflect it relates
  to the syslog mode (as opposed to sql/radius); for the same
  reasons, the accounting action is now called "acc_log_request"
- log_fmt allows now to specify what will be printed to syslog

auth module:
------------
- auth module has been split in auth, auth_db, auth_radius, group
  group_radius, uri and uri_radius 
- all the parameters that were part of former auth module are now 
  part of auth_db module
XXX 

im module:
----------
- im is no longer used and has been obsoleted by TM

exec module:
------------
- exec_uri and exec_user have been obsoleted by exec_dset; 
  exec_dset is identical to exec_uri in capabilities; it 
  additionaly passes content of request elements (header 
  fields and URI parts) in environment variables; users of 
  exec_user can use exec_dset now and use the "URI_USER"  
  variable to learn user part of URI
- exec_dset and exec_msg return false, if return value of 
  script does not euqal zero
- exec_dset takes an additional parameter, which enables 
  validation of SIP URIs returned by external application

jabber module:
--------------
- presence support for Jabber users is enabled loading the PA
  module and using subscribe("jabber") for SUBSCRIBE requests to
  jabber user 

msilo module:
-------------
- m_store has now a parameter to set what should be considered
  for storing as destination uri. This enables support for saving
  the messages on negative replies.

radius_acc module:
------------------
- radius_acc module has been removed and radius accounting 
  is now part of acc module

registrar/usrloc modules:
-------------------------
XXX
  
rr module:
----------
- addRecordRoute has been replaced with record_route
- rewriteFromRoute has been replaced with loose_route()
- a new option, "enable_full_lr" can be set to make life
  with misimplemented UAs easier and put LR in from "lr=on"

tm module:
----------
- t_reply_unsafe, used in former versions within reply_routes,
  is deprecated; now t_reply is used from any places in script
- t_on_negative is renamed to t_on_failure -- the old name just
  caused too much confusion
- FIFO t_uac used by some applications (like serweb) has been
  replaced with t_uac_dlg (which allows easier use by dialog-
  oriented applications, like click-to-dial) 
- if you wish to do forward to another destination from 
  failure_route (reply_route formerly), you need to call t_relay
  or t_relay_to explicitely now
- t_relay_to has been replaced with t_relay_to_udp and t_relay_to_tcp



obsolete/weather/weather/wm918/wx200d-1.2/wx200d.c
/*
 * @(#)$Id: wx200d.c,v 1.1 2002/09/23 19:12:51 bogdan Rel $
 *
 * Copyright (C) 1998 - 2002 Tim Witham <twitham@quiknet.com>
 *
 * (see the files README and COPYING for more details)
 *
 * This file implements the WX200 / WM918 data logger and server daemon.
 *
 */

#include <config.h>
#include <stdio.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif
#include <signal.h>
#include <sys/param.h>
#include <utime.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <syslog.h>
#include "wx200.h"
#include "serial.h"
#define POSTGRES	(defined(HAVE_LIBPQ) && defined(HAVE_LIBPQ_FE_H))
#if POSTGRES
#include "pg_api.h"
#endif

#ifdef MINDEBUG			/* write files every minute instead of day? */
#define MON(time)	(0)
#define DAY(time)	(time->tm_min)
#else
#define MON(time)	(time->tm_mon + 1)
#define DAY(time)	(time->tm_mday)
#endif

char error[80];
char *progname;
int *current = 0;		/* pointer to current socket handle */

#if POSTGRES
PGconn *pConn = NULL;		/* pointer to current postgres connection */
#endif

void
usage()
{
  printf("usage: %s [-h] [-d] [-a] [-b] [-z] [-p port] [-r] [-s device] [-w path]\n",
	 progname );
#if POSTGRES
  printf("\t[-g] [-c connection] [-t tablename] [-i interval]\n" );
#endif
  printf("  -h	show this help and exit
  -d	debug; don't fork to be a daemon
  -a	don't record tab-delimited ASCII data files
  -b	don't record abbreviated Binary data files
  -z	don't gZip daily data files after collection
  -p port	listen for client connections on port instead of %d
  -r		read a wireless WMR918 station instead of a wired WX200/WM918
  -s device	get WX200 data from device instead of %s
  -w path	write files in path instead of %s
", PORT, WX200, PACKAGE_DATA_DIR);
#if POSTGRES
  printf(
"  -g		enables recording values into a postgres database
  -c connection	postgres database connection instead of \n\t%s
  -t tablename	postgres table name instead of %s
  -i interval	postgres record interval instead of %d secs
", POSTGRES_CONN, POSTGRES_TABLE, POSTGRES_INTERVAL);
#endif

  exit(0);
}

int
stationbufread(int fd)
{
  if (wx200stationtype == 0)	/* original WX200/WM918 */
    return wx200bufread(fd);
  else if (wx200stationtype == 1) /* WMR918 by Dominique Le Foll */
    return wmr918bufread(fd);
  return -1;			/* huh? */
}

void
pipe_handler(int sig)		/* SIGPIPE signal handler */
{
  signal(SIGPIPE, SIG_IGN);
  if (sig == SIGPIPE) {		/* client went away */
    shutdown(*current, 2);
    close(*current);
    *current = -1;
  }
}

void
term_handler(int sig)		/* SIGTERM signal handler */
{
#if POSTGRES
  if (pConn)
    pg_finish( pConn );
#endif
  exit( 0 );
}

int
main(int argc, char **argv)
{
  int sd, pid, s, clen = sizeof(struct sockaddr_in), fd[CONNECTIONS];
  int *max = 0, afile = -1, bfile = -1, /* mfile = -1, */ c, first = 1;
  int pday = -1, pmon = -1, pwday = -1;
  int opt_a = 0, opt_b = 0, opt_d = 0, opt_z = 0, opt_p = PORT;
  int bufnum, i, j, bit, bits;
  time_t nowtime;
  struct tm *now;
  char *opt_s = NULL, *opt_w = PACKAGE_DATA_DIR;
  char afilename[128], bfilename[128], /* mfilename[128], */ command[256];
#if POSTGRES
  int opt_g = 0;
  char *opt_c = POSTGRES_CONN;
  char *opt_t = POSTGRES_TABLE;
  char *opt_i = NULL;
#endif
  struct sockaddr_in server, client;
  struct in_addr bind_address;
  fd_set rfds;
  struct timeval tv;
  struct utimbuf ut;
#if POSTGRES
  int nPgRet;			/* return of postgres insertion */
  time_t lasttime;		/* last time we made an insert */
  time_t pg_interval;		/* write to db every "pg_interval" seconds */
#endif

  wx200ignoreclock = 1;		/* computer clock overrides WX200's */
  progname = strrchr(argv[0], '/');
  if (progname == NULL)
    progname = argv[0];		/* global for error messages, usage */
  else
    progname++;
  while ((c = getopt(argc, argv, "Hhdrabzp:s:w:gc:t:i:")) != EOF) {
    switch(c) {
    case '?':
    case 'h':
    case 'H':			/* help */
      usage();
      break;
    case 'd':			/* do not fork and become a daemon */
      opt_d = 1;
      break;
    case 'a':			/* no tab-delimited ASCII log */
      opt_a = 1;
      break;
    case 'b':			/* no binary data log */
      opt_b = 1;
      break;
    case 'z':			/* do not gzip nightly after collection */
      opt_z = 1;
      break;
    case 'p':			/* port to use */
      opt_p = strtol(optarg, NULL, 0);
      break;
    case 's':			/* serial device to use */
      opt_s = optarg;
      break;
    case 'w':			/* path to write to */
      opt_w = optarg;
      break;
    case 'r':			/* select a WMR wireless radio station */
      wx200stationtype = 1;
      break;
#if POSTGRES
    case 'g':			/* enable postgres recording */
      opt_g = 1;
      break;
    case 'c':			/* postgres connection string */
      opt_c = optarg;
      break;
    case 't':			/* postgres tablename */
      opt_t = optarg;
      break;
    case 'i':			/* postgres recording time interval */
      opt_i = optarg;
      break;
#endif
    }
  }

  server.sin_family = AF_INET;
  bind_address.s_addr = htonl(INADDR_ANY);
  server.sin_addr = bind_address;
  server.sin_port = htons(opt_p);

  if ((s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)) == -1) {
    sprintf(error, "%s: socket", progname);
    perror(error);
    cleanup_serial(sd);
    exit(10);
  }
  {
    /* <dirkx@covalent.net> / April 2001 Minor change to allow quick
     * (re)start of deamon or client while there are pending
     * conncections during the quit. To avoid addresss/port in use
     * error.  */
    int i = 1;
    if (setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &i, sizeof(i)) == -1) {
      sprintf(error, "%s: setsockopt", progname);
      perror(error);
    }
  }
  if (bind(s, (struct sockaddr *)&server, sizeof(struct sockaddr_in)) == -1) {
    sprintf(error, "%s: bind", progname);
    perror(error);
    cleanup_serial(sd);
    exit(11);
  }
  if (listen(s, CONNECTIONS) == -1) {
    sprintf(error, "%s: listen", progname);
    perror(error);
    cleanup_serial(sd);
    exit(12);
  }

  /* If another server is running, the above will fail because of the
     busy port.  So no locking is needed here! */
  if ((sd = init_serial(opt_s)) < 0)
    exit(sd);

  umask(0022);
  wx200bufinit();		/* initialize all variables */
  wx200tabinit();
  for (i = 0; i < CONNECTIONS; i++) fd[i] = -1;
  tv.tv_sec = 0;
  tv.tv_usec = 0;

  openlog(progname, LOG_PID | (opt_d ? LOG_PERROR : 0), LOG_LOCAL4);

  if (!opt_d) {			/* setup has worked; now become a daemon? */
    if ((pid = fork()) == -1) {
      syslog(LOG_ERR, "can't fork() to become daemon: %m");
      cleanup_serial(sd);
      exit(20);
    } else if (pid)
      exit (0);
    setsid();
    for (i = 0; i < NOFILE; i++)
      if (i != sd && i != s) close(i);
  }

  /* catch signals to close the database connection */
  signal( SIGTERM, term_handler );/* termination */
  signal( SIGPWR, term_handler ); /* power failure */

#if POSTGRES
  if( opt_g ) {			/* postgres recording enabled */
    if( opt_i != NULL )
      pg_interval = atol( opt_i );
    else
      pg_interval = POSTGRES_INTERVAL;
    lasttime = 0;
    pConn = pg_init( opt_c, opt_t );
  }
#endif

  while (1) {			/* live forever, listening and logging */
    FD_ZERO(&rfds);
    FD_SET(s, &rfds);
    if (select(s + 1, &rfds, NULL, NULL, &tv) > 0) {
      for (current = fd;
	   (*current > 0) && (current < fd + CONNECTIONS - 1); current++);
      if (current > max) max = current;
      if ((*current = accept(s, (struct sockaddr *)&client, &clen)) != -1)
	for (i = 0; i < GROUPS; i++) { /* give current data to new client */
	  write(*current, groupbuf[i], grouplen[i]);
	}
    }
    FD_ZERO(&rfds);
    FD_SET(sd, &rfds);
    while ((select(sd + 1, &rfds, NULL, NULL, &tv) > 0)
	   && ((bufnum = stationbufread(sd)) >= 0)) {
      switch (bufnum) {
      case GROUPF:		/* timestamp? check for new day */
      case GROUP1:
      case GROUP2:
      case GROUP3:
      case GROUP4:		/* 1-7 is a WMR918 multi-group (3 bits) */
      case GROUP5:
      case GROUP6:
      case GROUP7:
      case GROUP8:
	nowtime = time(NULL);
	now = localtime(&nowtime);
	if ((DAY(now) != pday) || first) {
	  if (!first) {		/* non-first files? close current ones */
	    if (bfile > -1) close(bfile);
	    if (afile > -1) close(afile);
	    ut.actime = ut.modtime = time(NULL); /* make sure .tab is as new */
	    utime(afilename, &ut);
	    if (!opt_z) {	/* compress yesterday's files */
	      if (bfile > -1 || afile > -1) {
		sprintf(command, "gzip %s %s &", bfile > -1 ? bfilename : "",
			afile > -1 ? afilename : "");
		system(command);
	      }
	    }
	  }
	  sprintf(bfilename, "%s/%02d%02d%02d.bin", opt_w,
		  1900 + now->tm_year, MON(now), DAY(now));
	  sprintf(afilename, "%s/%02d%02d%02d.tab", opt_w,
		  1900 + now->tm_year, MON(now), DAY(now));
	  /* 	  sprintf(mfilename, "%s/%02d%02d%02d.mem", opt_w, */
	  /* 		  1900 + now->tm_year, MON(now), DAY(now)); */
	  if (!opt_b &&
	      (bfile = open(bfilename, O_WRONLY|O_CREAT|O_APPEND, 00644)) < 0) {
	    syslog(LOG_ERR, "can't write %s: %m", bfilename);
	    bfile = -1;
	  }
	  if (!opt_a &&
	      (afile = open(afilename, O_WRONLY|O_CREAT|O_APPEND, 00644)) < 0) {
	    syslog(LOG_ERR, "can't write %s: %m", bfilename);
	    afile = -1;
	  }
	  if (!first) {		/* non-first file?  begin w/ current data */
	    if (bfile > -1)
	      for (i = 0; i < GROUPS; i++) {
		write(bfile, groupbuf[i], grouplen[i]);
	      }
	  }
	  wx200bufinit();	/* forget memory to re-get all groups */
	  wx200tabinit();
	  wx200parse(-1);	/* reset appropriate virtual memories to now */
	  pday = DAY(now);
	  pmon = MON(now);
	  pwday = now->tm_wday;
	  first = 0;
	}
      case GROUP9:
      case GROUPA:
      case GROUPB:
      case GROUPC:
	if (bufnum < GROUP8) {	/* WMR918 tweaks up to 3 groups at once */
	  bits = bufnum;
	  bufnum = GROUP8;
	  j = GROUPA;
	} else {		/* otherwise, just log this one group */
	  bits = 0x01;
	  j = bufnum;
	}
	for (i = bit = 0; bufnum <= j; bufnum++) {
	  if (bits & (0x01 << bit)) { /* this group changed? */
	    if (wx200tab(bufnum)) /* calls wx200parse(bufnum) */
	      i = 1;
	    if (bfile > -1)
	      write(bfile, groupbuf[bufnum], grouplen[bufnum]);
	    for (current = fd; current <= max; current++)
	      if (*current > 0) {	/* active socket connection */
		signal(SIGPIPE, pipe_handler);
		write(*current, groupbuf[bufnum], grouplen[bufnum]);
	      }
	  }
	  bit++;
	}
	if (afile > -1 && i)
	  write(afile, tabbuf, strlen(tabbuf));
#if POSTGRES
	if( opt_g ){		/* postgres recording enabled */
	  if( nowtime - lasttime > pg_interval ) {
	    lasttime = nowtime;
	    nPgRet = pg_insert( pConn, opt_t, &wx, 1 );
	    if( nPgRet == -2 ) { /* retrying the connection */
	      pConn = pg_init( opt_c, opt_t );
	    }
	  }
	}
#endif
      }
    }
    sleep(1);			/* no hurry, be a nice daemon */
  }
  cleanup_serial(sd);
  exit(0);
}
